### PR TITLE
Add docs

### PR DESCRIPTION
**Pull request by solidgoldpig**
First pass.

Still some holes - but once you've identified them and I've (hopefully) explained them, you can fill them in

### CODE REVIEW

File: documentation/running/runner/component.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+- [Checkboxes](/component/checkboxes)
+- [Date](/component/date)
+- [Number](/component/number)
+- [Radios](/component/radios)
+- [Select](/component/select)
+- [Text](/component/text)
+- [Textarea](/component/textarea)
+
+NB. File upload is currently not supported.
```
Comment (umaar) [inline documentation/running/runner/component.md]: Is this still correct?

File: documentation/running/runner/flow.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+The flow through the service is determined by:
+
+- starting at the entry point
+- moving through the pages that are referenced as its steps
+- recursing into any steps encountered on any of those pages
+- looping over multiple page instances
+- short-circuiting this process if an explicit nextPage is defined 
+
+*TODO: Include explicit nextPage in diagram, update labels*
```
Comment (umaar) [inline documentation/running/runner/flow.md]: Reminder

File: bin/adr-generate-toc [hunk 3]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env node
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply bin/adr-generate-toc]: https://github.com/ministryofjustice/form-builder/blob/add-docs/tools.md ;)

File: documentation/publishing/publisher--functions.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+### Deploy a new service
+- Deploy a built service docker image:tag from the service container repository to the Hosting
+- Configure routing of URL to the service
+  - Inputs
+    - Service image id:tag
+  - Outputs
+    - A running service
+
+### Deploy a specific (newer/older) version of an existing service
+- Deploy a built service docker image:tag from the service container repository to the Hosting
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: Copied verbatim - one for the publisher gang to sort in the fullness of time

File: documentation/publishing/publisher--functions.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    - Service JSON git repo:commit
+    - Service config / env vars
+    - SCR URL
+    - SCR creds
+  - Outputs
+    - Tagged docker image ID for the service
+
+### Deploy a new service
+- Deploy a built service docker image:tag from the service container repository to the Hosting
+- Configure routing of URL to the service
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: As above

File: documentation/publishing/publisher--functions.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
+# FB Publisher functionality
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: As above

File: documentation/running/runner/component.md (hunk 1 above)
Comment (solidgoldpig) [reply documentation/running/runner/component.md]: Nope - fixed

File: documentation/building/building.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+Specifications for components and pages are defined as [JSON schemas](../specifications/schemas.md)
+
+## Core components
+
+The core components are defined in [fb-components-core](https://github.com/ministryofjustice/fb-components-core)
+
+## Editor
+
+The [Editor](editor/editor.md) is an app that provides a UI to build a forms (ie. create the necessary data) using the components specified for the form.
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/building/building.md]: Sorted

File: documentation/building/editor/editor.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Form Builder Editor
+
+The editor is an app that provides a UI to build a forms (ie. create the necessary data) using the components specified for the form.
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/building/editor/editor.md]: Done

File: documentation/platform/error-responses.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# Error responses
+
+All error responses should be sent as `json` and take the form
+
+```json
+{
+  "code": {number},
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/error-responses.md]: Done

File: documentation/platform/file-upload/user-filestore--api.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    "expires": "<integer> days"
+  }
+}
+```
+
+###  A. As per "Check request correctly signed and meets requirements"
+
+###  B. Check file
+
+- Perform size check if `policy.max_size` is present
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/file-upload/user-filestore--api.md]: C'est moi 😊

File: documentation/platform/logging.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+Draft - include RAILS_LOG ENV var info
+
+(Better still, turn it on by default in non-prod)
+
+## Aggregated logs
+
+Logs that have been exposed to the container they run in's stdout can be accessed via Kibana
+
+https://kibana.apps.cloud-platform-live-0.k8s.integration.dsd.io
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/logging.md]: Raised an issue for this

File: documentation/platform/logging.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+Forms can be identified by setting filtering by `kubernetes.container_name`, setting its value to that of the form you're after
+
+eg.
+
+<a href="https://kibana.apps.cloud-platform-live-0.k8s.integration.dsd.io/_plugin/kibana/app/kibana#/discover?_g=(refreshInterval:(value:5000),time:(from:now-15m,mode:quick,to:now))&_a=(columns:!(log),filters:!((query:(match:(kubernetes.container_name:(query:moj-leavers-form,type:phrase))))),interval:auto,query:(language:lucene,query:''),sort:!('@timestamp',desc))">Link to moj-leavers-form last 15 minutes</a>
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/logging.md]: As above - issue raised

File: documentation/platform/logging.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
+# Logging
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/logging.md]: Added issue

File: documentation/platform/network-security.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+  - access only allowed to `fb-submitter-api` and `fb-submitter-workers`
+
+## formbuilder-services-{platformEnv}-{deploymentEnv}
+
+- ServiceAccounts
+  - `fb-publisher-workers-{platformEnv}` in `formbuilder-publisher-{platformEnv}`namespace - allows deployments
+  - `formbuilder-service-token-cache-{platformEnv}-{deploymentEnv}` in `formbuilder-platform-{platformEnv}-{deploymentEnv}` namespace - allows service tokens to be read
+- In-bound
+  - citizen
+    - WHAT DOES THE COOKIE LOOK LIKE?
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/network-security.md]: Added issue

File: documentation/platform/user-data/user-datastore--api.md [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+## API
+
+### No Ingress, HTTPS-only
+
+Ideally this would be achieved with a Service Mesh (e.g. https://istio.io) but this is not provided by the Cloud Platform at the time of writing, nor is it currently on their near-term roadmap.
+Therefore we’ll experiment with an internal load balancer model (see https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer and https://kubernetes.io/docs/concepts/services-networking/service/#ssl-support-on-aws )
+
+*UPDATE 10/10/2018*  
+*Cloud Platform team are now discouraging the use of Kubernetes services of type ‘LoadBalancer’, have accepted that secure internal comms is a requirement, and now have a card for this on their roadmap - as a result, considering we are not planning to hosting any live public-facign services in the immediate future, we’ll hold on for whatever solution they end up with.*
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--api.md]: The update is still very relevant - it addresses what Phil was talking about re: JWT

File: documentation/platform/platform--environments.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+Apps within platform apps are shared by all the forms in the corresponding service apps namespace
+
+[![Form Builder environments overview](images/fb-platform-environment-overview.png)](images/fb-platform-environment-overview.png)
+
+([Source file (Axure)](files/fb-platform-environment-overview.rp))
+
+### Per Branch? Per commit?
+
+TODO: fill in
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/platform--environments.md]: Added issue

File: documentation/platform/platform--overview.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
+# Form Builder overview
+
+TODO: this probably isn't necessary
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/platform--overview.md]: We can probably delete this file - raised issue

File: documentation/platform/service-token-cache.md [hunk 18]
```diff
@@ -0,0 +1,3 @@
+# Service token cache
+
+TODO: Explain this
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/service-token-cache.md]: Added issue

File: documentation/platform/user-data/user-datastore--api.md [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+## Introduction
+
+Services written & deployed with FormBuilder require a component to store data submitted by their users in a secure way. We have agreed a suitable implementation with the Information Security team, discussed in detail in "Runner / User data store Threats & Mitigations".
+
+This document summarises requirements for the User Data Store component & knock-on implications for the other components affected - the Runner & Publisher. 
+
+
+## System Context
+
+TODO: Insert image
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--api.md]: Image already added ;)

File: documentation/platform/user-data/user-datastore--api.md [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+Ideally this would be achieved with a Service Mesh (e.g. https://istio.io) but this is not provided by the Cloud Platform at the time of writing, nor is it currently on their near-term roadmap.
+Therefore we’ll experiment with an internal load balancer model (see https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer and https://kubernetes.io/docs/concepts/services-networking/service/#ssl-support-on-aws )
+
+*UPDATE 10/10/2018*  
+*Cloud Platform team are now discouraging the use of Kubernetes services of type ‘LoadBalancer’, have accepted that secure internal comms is a requirement, and now have a card for this on their roadmap - as a result, considering we are not planning to hosting any live public-facign services in the immediate future, we’ll hold on for whatever solution they end up with.*
+
+### RDS Postgres Database
+
+DynamoDB may suit the use case better in the long-term, but is not yet supported by Cloud Platforms.
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--api.md]: Raised issue

File: documentation/platform/user-data/user-datastore--api.md [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+  - If the token is not present, the response will be 401 Unauthorised
+  - If the token is not valid, the response will be 403 Forbidden
+
+
+
+## Publisher
+
+Publisher needs:
+
+- A field adding to the Service record to store the service token
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--api.md]: Covered by issue raised above

File: documentation/platform/user-data/user-datastore--api.md [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+- Auto-generation of the service token on creation
+- A button in the UI for generating a new service token
+- UI to allow inspection of the service token (treated like a password field - i.e. obfuscated by default, only revealed when explicitly needed)
+- SERVICE_TOKEN to be injected as an environment variable when deploying a service
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--api.md]: That would be great - if the documentation for publisher was in a fit state. But, yes, we should. When it exists.

File: documentation/platform/user-data/user-datastore--threats.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
+# User Datastore - Threats and mitigation
+
+Users will be submitting potentially many different pieces of potentially very sensitive information to many services deployed on Form Builder. This data will not be permanently stored, but must be stored temporarily to allow a user to complete their journey.
+
+Such a large cross-section of personal information must be stored and managed carefully. This document outlines the proposed model to mitigate each identified risk.
+
+
+TODO: Insert image
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--threats.md]: Raised issue

File: documentation/platform/user-data/user-datastore--threats.md [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+## Network Segmentation / Isolation
+
+The internet-facing Runner services will not have direct access to the datastore, only via API calls. The API will exist in a separate namespace from the Runner services, and only the API pods will have access to the backing database.
+
+## Security in transit
+
+We will use TLS for communication between the Runner and the datastore. Implementation mechanism TBD, but we will experiment with an internal load balancer model (see https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer and https://kubernetes.io/docs/concepts/services-networking/service/#ssl-support-on-aws ) until Cloud Platforms provide a service mesh (e.g. https://istio.io  )
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/platform/user-data/user-datastore--threats.md]: As above - issue raised

File: documentation/publishing/publisher--data.md [hunk 25]
```diff
@@ -0,0 +1,35 @@
+# Publisher data
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--data.md]: It's the only publisher documentatiomn at the moment. As above - issue raised

File: documentation/publishing/publisher--data.md [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+## Users
+
+- Name
+- Email
+- Timestamps
+
+## Teams
+
+- Name
+- Email(?)
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--data.md]: As above - issue raised

File: documentation/publishing/publisher--functions.md [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+
+##  Deployment functions
+
+### Build service
+- Build a docker image for the given service, push it to the service container repository (SCR)
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: Yep - but covered by issue to review/write proper docs for publisher

File: documentation/publishing/publisher--functions.md [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+  - Inputs
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: As above - issue raised

File: documentation/running/runner/basics-example-service.md [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+
+- a start page
+- a page that asks the user for a greeting to address the planet with
+- a page that asks the user for their favourite planet
+- a confirmation page that replays the user’s answers
+
+
+## Service pages
+
+## Start page - `pageStartExample.json`
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/running/runner/basics-example-service.md]: Fixed

File: documentation/running/runner/about-data.md [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+  - how to render the [Editor](/process/editor) UI
+  - how and where to deploy the service
+  - what other services to interact with
+  - how to render the citizen-facing service
+  - how to validate [user data](#user-data)
+  - where to send the user data at the completion of the journey
+
+This data is made up of [schemas](/glossary#schema), test [instances](/glossary#instance), documentation and templates and are stored in [specification modules](storage#specification-data)
+
+### Analytics data
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/running/runner/about-data.md]: Added issue

File: documentation/publishing/publishing.md [hunk 31]
```diff
@@ -0,0 +1,18 @@
+# Publishing forms with Form Builder
+
+Heroku-like, deploys apps to MoJ’s Cloud Platform
+
+TODO: more here
+
+TODO: Explain CONFIG PARAMS
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publishing.md]: Added issue

File: documentation/publishing/publishing.md [hunk 32]
```diff
@@ -0,0 +1,18 @@
+# Publishing forms with Form Builder
+
+Heroku-like, deploys apps to MoJ’s Cloud Platform
+
+TODO: more here
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publishing.md]: Added issue

File: documentation/publishing/publisher--functions.md [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+- Deploy a built service docker image:tag from the service container repository to the Hosting
+- Configure routing of URL to the service
+  - Inputs
+    - Service image id:tag
+  - Outputs
+    - A running service
+
+### Deploy a specific (newer/older) version of an existing service
+- Deploy a built service docker image:tag from the service container repository to the Hosting
+- Manage smooth transition between previous version and newly-deployed version
```
[the remark this answers is not in the document]
Comment (solidgoldpig) [reply documentation/publishing/publisher--functions.md]: I don't know - added issue about writing proper docs for publisher

**Pull request merged**